### ISSUE TITLE
Facing issue while resolving dependency in rpm installation

### ISSUE THREAD

**Issue by tapakund** [openSUSE/libsolv#406]
While trying to install tuned in photon OS 3.0, it's forcing user to switch to linux-aws, instead of upgrading the linux package to the latest. The version of linux installed is linux-4.19.132-6.ph3.x86_64 in the machine.

Also, it's trying to install linux-aws, linux-python3-perf and linux-tools as dependencies, all at version 4.19.145-4.ph3. However, the latest available versions of those packages in the repo is actually 4.19.148-3.ph3

So, here is how the package dependency goes:

tuned requires linux-python3-perf.
linux-python3-perf requires linux-tools.
linux-tools requires any of the linux flavour rpm: 
```
%package tools
Summary:        This package contains the 'perf' performance analysis tools for Linux kernel
Group:          System/Tools
Requires:       (%{name} = %{version} or linux-esx = %{version} or linux-aws = %{version} or linux-rt = %{version})
```

So, we have explored different scenarios:

We kept linux-4.19.145-4, linux-aws-4.19.145-4, linux-tools-4.19.145-4, linux-python3-perf-4.19.145-4 and linux-4.19.148-5, linux-aws-4.19.148-5, linux-tools-4.19.148-5, linux-python3-perf-4.19.148-5 in a test-repo.
 
```
root@photon-945c894e227f [ /test-rpms ]# ls | grep linux
linux-4.19.145-4.ph3.x86_64.rpm
linux-4.19.148-5.ph3.x86_64.rpm
linux-aws-4.19.145-4.ph3.x86_64.rpm
linux-aws-4.19.148-5.ph3.x86_64.rpm
linux-python3-perf-4.19.145-4.ph3.x86_64.rpm
linux-python3-perf-4.19.148-5.ph3.x86_64.rpm
linux-tools-4.19.145-4.ph3.x86_64.rpm
linux-tools-4.19.148-5.ph3.x86_64.rpm
python3-linux-procfs-0.6.1-1.ph3.noarch.rpm
root@photon-945c894e227f [ /test-rpms ]#
```

Trying to install tuned, its pulling in linux-aws-4.19.145-4 even though we have linux-aws-4.19.148-5 or even linux-4.19.148-5.

```
root@photon-945c894e227f [ /dnf/dnf ]# tdnf install tuned
 
Installing:
**linux-aws                                             x86_64                     4.19.145-4.ph3                      tuned-test                   19.58M 20527390**
linux-tools                                           x86_64                     4.19.145-4.ph3                      tuned-test                   31.21M 32731264
linux-python3-perf                                    x86_64                     4.19.145-4.ph3                      tuned-test                    354.79k 363309
tuned                                                 noarch                     2.13.0-3.ph3                        tuned-test                    447.86k 458612
 
Total installed size:  51.58M 54080575
Is this ok [y/N]: n
Error(1032) : Operation aborted.
root@photon-945c894e227f [ /dnf/dnf ]#
```
```
root@photon-945c894e227f [ /libsolv-debug/debugdata ]# tdnf list linux-aws
linux-aws.x86_64                                                                                 4.19.145-4.ph3                                              tuned-test
linux-aws.x86_64                                                                                 4.19.148-5.ph3                                              tuned-test
root@photon-945c894e227f [ /libsolv-debug/debugdata ]#
```
 
After that, we removed linux-4.19.145-4, linux-aws-4.19.145-4, linux-tools-4.19.145-4 and linux-python3-perf-4.19.145-4 from the test repo and replaced with linux-4.19.138-1, linux-aws-4.19.138-1, linux-tools-4.19.138-1 and linux-python3-perf-4.19.138-1.
 
```
root@photon-945c894e227f [ /test-rpms ]# ls | grep linux
linux-4.19.138-1.ph3.x86_64.rpm
linux-4.19.148-5.ph3.x86_64.rpm
linux-aws-4.19.138-1.ph3.x86_64.rpm
linux-aws-4.19.148-5.ph3.x86_64.rpm
linux-python3-perf-4.19.138-1.ph3.x86_64.rpm
linux-python3-perf-4.19.148-5.ph3.x86_64.rpm
linux-tools-4.19.138-1.ph3.x86_64.rpm
linux-tools-4.19.148-5.ph3.x86_64.rpm
python3-linux-procfs-0.6.1-1.ph3.noarch.rpm
root@photon-945c894e227f [ /test-rpms ]#
```
With this, I see tdnf is updating linux to the latest version i.e., 4.19.148-5 and installing tuned.

```
root@photon-945c894e227f [ /dnf/dnf ]# tdnf install tuned
 
Installing:
linux-tools                                           x86_64                     4.19.148-5.ph3                      tuned-test                   31.22M 32734291
linux-python3-perf                                    x86_64                     4.19.148-5.ph3                      tuned-test                    354.79k 363309
tuned                                                 noarch                     2.13.0-3.ph3                        tuned-test                    447.86k 458612
 
Total installed size:  32.00M 33556212
 
Upgrading:
**linux                                                 x86_64                     4.19.148-5.ph3                      tuned-test                   26.22M 27495988**
 
Total installed size:  26.22M 27495988
Is this ok [y/N]: n
Error(1032) : Operation aborted.
root@photon-945c894e227f [ /dnf/dnf ]#
```

Later, we built linux-4.19.145-4, linux-aws-4.19.145-4, linux-tools-4.19.145-4 and linux-python3-perf-4.19.145-4 locally as well and copied that in test repo.
 
We still see same behaviour, it’s still forcing to install linux-aws-4.19.145-4.
 
```
root@photon-945c894e227f [ /test-rpms ]# ls -ltr | grep linux
-rw-r----- 1 root root    41513 Jun 23 06:56 python3-linux-procfs-0.6.1-1.ph3.noarch.rpm
-rw-r----- 1 root root 10237132 Aug 18 06:23 linux-tools-4.19.138-1.ph3.x86_64.rpm
-rw-r----- 1 root root   160561 Aug 18 06:23 linux-python3-perf-4.19.138-1.ph3.x86_64.rpm
-rw-r----- 1 root root 16989511 Aug 18 06:24 linux-aws-4.19.138-1.ph3.x86_64.rpm
-rw-r----- 1 root root 22499160 Aug 18 06:24 linux-4.19.138-1.ph3.x86_64.rpm
-rw-r----- 1 root root 22539015 Oct 18 10:16 linux-4.19.148-5.ph3.x86_64.rpm
-rw-r----- 1 root root   161584 Oct 18 10:16 linux-python3-perf-4.19.148-5.ph3.x86_64.rpm
-rw-r----- 1 root root 16996401 Oct 18 10:16 linux-aws-4.19.148-5.ph3.x86_64.rpm
-rw-r----- 1 root root 10213415 Oct 18 10:16 linux-tools-4.19.148-5.ph3.x86_64.rpm
-rw-r----- 1 root root 16989779 Oct 21 12:56 linux-aws-4.19.145-4.ph3.x86_64.rpm
-rw-r----- 1 root root 22538640 Oct 21 12:56 linux-4.19.145-4.ph3.x86_64.rpm
-rw-r----- 1 root root 10212969 Oct 21 12:56 linux-tools-4.19.145-4.ph3.x86_64.rpm
-rw-r----- 1 root root   160830 Oct 21 12:56 linux-python3-perf-4.19.145-4.ph3.x86_64.rpm
root@photon-945c894e227f [ /test-rpms ]#
 
root@photon-945c894e227f [ /dnf/dnf ]# tdnf install tuned
Refreshing metadata for: 'VMWare Photon Linux ISO 3.0 (x86_64)'
tuned-test                              228825   100%
Installing:
**linux-aws                                             x86_64                     4.19.145-4.ph3                      tuned-test                   19.58M 20527278**
linux-tools                                           x86_64                     4.19.145-4.ph3                      tuned-test                   31.21M 32731272
linux-python3-perf                                    x86_64                     4.19.145-4.ph3                      tuned-test                    354.79k 363309
tuned                                                 noarch                     2.13.0-3.ph3                        tuned-test                    447.86k 458612
 
Total installed size:  51.58M 54080471
Is this ok [y/N]: n
Error(1032) : Operation aborted.
root@photon-945c894e227f [ /dnf/dnf ]#
 ```

If we just try to install linux-python3-perf or linux-tools, I do not see any issue. Its pulling in latest linux version along with that.

```
root@photon-945c894e227f [ /libsolv-debug ]# tdnf install  linux-python3-perf

Installing:
linux-tools                                           x86_64                     4.19.148-5.ph3                      tuned-test                   31.22M 32734291
linux-python3-perf                                    x86_64                     4.19.148-5.ph3                      tuned-test                    354.79k 363309

Total installed size:  31.56M 33097600

Upgrading:
linux                                                 x86_64                     4.19.148-5.ph3                      tuned-test                   26.22M 27495988

Total installed size:  26.22M 27495988
Is this ok [y/N]: n
Error(1032) : Operation aborted.
root@photon-945c894e227f [ /libsolv-debug ]#
```

```
root@photon-945c894e227f [ /libsolv-debug ]# tdnf install  linux-tools

Installing:
linux-tools                                           x86_64                     4.19.148-5.ph3                      tuned-test                   31.22M 32734291

Total installed size:  31.22M 32734291

Upgrading:
linux                                                 x86_64                     4.19.148-5.ph3                      tuned-test                   26.22M 27495988

Total installed size:  26.22M 27495988
Is this ok [y/N]: n
Error(1032) : Operation aborted.
root@photon-945c894e227f [ /libsolv-debug ]#
```

We are still figuring out to root cause this behaviour on why this is forcing to install linux-aws and that too an older version. Have collected the debugdata and attached the tarball here if it helps in any way to find out the reason behind this behaviour.

We have verified the same with dnf as well and see the same behaviour.

We have 0.6.35 version of libsolv installed in the machine.

```
root@photon-945c894e227f [ /libsolv-debug ]# rpm -q libsolv
libsolv-0.6.35-3.ph3.x86_64
root@photon-945c894e227f [ /libsolv-debug ]#
```

Looking for help.

Thanks!
[debugdata.tar.gz](https://github.com/openSUSE/libsolv/files/5428991/debugdata.tar.gz)

**Comment by tapakund** [comment]
Can someone suggest a resolution for this issue?

**Comment by mlschroe** [comment]
This happens because the solver tries to fit the packages to the installed ones. There is actually a countermeasure in the code (the so-called "choice rules") to prevent the installation of new packages if a dependency can also be met by updating a package, but for some reason it does not work in this case.

I'll try to change the choice rule generation code.

**Comment by tapakund** [comment]
Thanks @mlschroe for looking into this. Let me know if you are aware of a workaround to overcome this issue for time being?

**Comment by mlschroe** [comment]
Ok, now fixed with commit 958c8396ade46920b87f29f1f75456d97cc63812

**Comment by tapakund** [comment]
Thanks!
